### DISCUSSION TITLE
Workflow Unit Tests

### DISCUSSION THREAD

**Discussion by srblackie** [square/workflow-kotlin#340] [Q&A] [answered]
What's the preferred way to write Workflow Unit Tests?

**Comment by rjrjr** [comment, marked as answer]
If you're asking how to choose between `Workflow.launchForTestingFromStartWith` and `Workflow.testRender`, the latter allows for more focused tests, and is probably the best choice for unit tests. (We gave it the more concise name as an under-documented hint in that direction.) It allows you to test a single rendering at a time.

`launchForTestingFromStartWith` is better suited for integration tests, e.g. to test an entire workflow tree. It can be a decent alternative to an espresso test.
